src/defrag.c

```diff
@@ -844,76 +882,434 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-                         server.stat_active_defrag_scanned - prev_scanned > 64)) {
-                if (quit || ustime() > endtime) {
-                    if (key_defragged != server.stat_active_defrag_hits)
-                        server.stat_active_defrag_key_hits++;
-                    else
-                        server.stat_active_defrag_key_misses++;
-                    return 1;
-                }
-                iterations = 0;
-                prev_defragged = server.stat_active_defrag_hits;
-                prev_scanned = server.stat_active_defrag_scanned;
+            state.cursor = kvstoreDictLUTDefrag(kvs, state.cursor, dictDefragTables);
+            if (getMonotonicUs() >= endtime) return false;
+        } while (state.cursor != 0);
+        state.slot = -1; // Done with defrag of main structures
+    }
+
+    while (true) {
+        if (++iterations > 16 || server.stat_active_defrag_hits - prev_defragged > 512 || server.stat_active_defrag_scanned - prev_scanned > 64) {
+            if (getMonotonicUs() >= endtime) break;
+            iterations = 0;
+            prev_defragged = server.stat_active_defrag_hits;
+            prev_scanned = server.stat_active_defrag_scanned;
+        }
+
+        if (precontinue_fn) {
+            if (privdata) *(kvstoreIterState *)privdata = state;
+            if (!precontinue_fn(endtime, privdata)) return false;
+        }
+
+        if (!state.cursor) {
+            // If there's no cursor, we're ready to begin a new kvstore slot.
+            if (state.slot == -1) {
+                state.slot = kvstoreGetFirstNonEmptyDictIndex(kvs);
+            } else {
+                state.slot = kvstoreGetNextNonEmptyDictIndex(kvs, state.slot);
             }
-        } while (defrag_later_cursor);
-        if (key_defragged != server.stat_active_defrag_hits)
-            server.stat_active_defrag_key_hits++;
-        else
-            server.stat_active_defrag_key_misses++;
-    } while (1);
+
+            if (state.slot == -1) return true; // ALL DONE!
+        }
+
+        // Whatever privdata's actual type, this function requires that it begins with kvstoreIterState.
+        if (privdata) *(kvstoreIterState *)privdata = state;
+        state.cursor = kvstoreDictScanDefrag(kvs, state.slot, state.cursor,
+                                             scan_fn, defragfns, privdata);
+    }
+
+    return false; // Not done yet
+}
+
+
+// Note: target is a DB, (not a KVS like most stages)
+static bool defragStageDbKeys(monotime endtime, void *target, void *privdata) {
+    UNUSED(privdata);
+    serverDb *db = (serverDb *)target;
+
+    static defragKeysCtx ctx; // STATIC - this persists
+    if (endtime == 0) {
+        ctx.db = db;
+        // Don't return yet.  Call the helper with endtime==0 below.
+    }
+    serverAssert(ctx.db == db);
+
+    // Note: for DB keys, we use the start/finish callback to fix an expires table entry if
+    //       after the main DB entry is moved.
```

Comment:
```suggestion
    // Note: for DB keys, we use the start/finish callback to fix an expires table entry if
    //       the main DB entry is moved.
```

Feels like an extra word, after could also work too, but if after doesn't make sense.